lib/ddtrace/contrib/http/instrumentation.rb

```diff
@@ -63,49 +72,57 @@ def request(req, body = nil, &block) # :yield: +response+
             end
           end
 
-          def annotate_span!(span, request, response)
+          def annotate_span!(span, request, response, request_options)
             span.set_tag(Datadog::Ext::HTTP::URL, request.path)
             span.set_tag(Datadog::Ext::HTTP::METHOD, request.method)
             span.set_tag(Datadog::Ext::HTTP::STATUS_CODE, response.code)
 
-            if request.respond_to?(:uri) && request.uri
-              span.set_tag(Datadog::Ext::NET::TARGET_HOST, request.uri.host)
-              span.set_tag(Datadog::Ext::NET::TARGET_PORT, request.uri.port.to_s)
-            else
-              span.set_tag(Datadog::Ext::NET::TARGET_HOST, @address)
-              span.set_tag(Datadog::Ext::NET::TARGET_PORT, @port.to_s)
-            end
+            host, port = host_and_port(request)
+            span.set_tag(Datadog::Ext::NET::TARGET_HOST, host)
+            span.set_tag(Datadog::Ext::NET::TARGET_PORT, port.to_s)
 
-            # Set analytics sample rate
-            Contrib::Analytics.set_sample_rate(span, analytics_sample_rate) if analytics_enabled?
+            set_analytics_sample_rate(span, request_options)
 
             case response.code.to_i
             when 400...599
               span.set_error(response)
             end
           end
 
-          def datadog_pin
+          def set_analytics_sample_rate(span, request_options)
+            return unless analytics_enabled?(request_options)
+            Contrib::Analytics.set_sample_rate(span, analytics_sample_rate(request_options))
+          end
+
+          def datadog_pin(config = Datadog.configuration[:http])
             @datadog_pin ||= begin
-              service = Datadog.configuration[:http][:service_name]
-              tracer = Datadog.configuration[:http][:tracer]
+              service = config[:service_name]
+              tracer = config[:tracer]
```

Comment:
There's a subtle bug here in that someone could call `datadog_pin` on this object with no args before running a request. It would then use the `:default` configuration to supply args to the pin. If the request is then run for a hostname which matches a regex configuration, we'll use the wrong configuration since it's been memoized.

I didn't have enough context to know if this is actually a problem, or what people use custom service names on pins for, or if the pin should just be removed since they look deprecated, or ...

Reply:
Yeah I think this could be a problem, given that someone might build an HTTP object, set something on its `datadog_pin` then call it with a request resulting in this scenario.

I think that's one of the tricky things about this "by hostname" configuration: we have 1) default, 2) "by host" and 3) by object configuration which all has to be merged together to apply the settings that make the most sense for usability. Inevitably I think this resolution has to happen at request time (less than ideal, but unavoidable.)